File: 백준/Silver/11047. 동전 0/동전 0.py
from sys import stdin
def findKey(n, k, lst):
    for i in range(1, n):
        if k>=lst[0]:
            return i-1
        if k < lst[i - 1] and k >= lst[i]:
            return i

    return n - 1


lst = []
cnt = 0

n, k = map(int, stdin.readline().split())

for _ in range(n):
    lst.append(int(input()))

lst.reverse()

while k != 0:
    f = findKey(n, k, lst)
    cnt += k // lst[f]
    k %= lst[f]

print(cnt)

File: 백준/Silver/11047. 동전 0/동전 0.java
import java.util.Collections;
import java.util.ArrayList;
import java.util.Scanner;

public class Main {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        int k = scanner.nextInt();
        int cnt=0, f;
        ArrayList<Integer> lst = new ArrayList<Integer>();
        for (int i=0; i<n; i++)
            lst.add(scanner.nextInt());
        Collections.sort(lst, Collections.reverseOrder()); // descending sort

        while (k!=0) {
            f=findKey(n, k, lst);
            cnt+=k/lst.get(f);
            k%=lst.get(f);
        }
        System.out.println(cnt);
    }
    public static int findKey(int n, int k, ArrayList<Integer> lst) {
        for (int i=1; i<n; i++) {
            if (k>= lst.get(0)) {
                return --i;
            }
            if (k<lst.get(i-1) && k>=lst.get(i))
                return i;
        }
        return n-1;
    }
}
